### DISCUSSION TITLE
How to use `get_val()` / `set_val()`

### DISCUSSION THREAD

**Discussion by msmftc** [fvutils/pyvsc#124] [Q&A]
@mballance or others,

I want to use the `get_val()` and `set_val()` methods that are provided for scalar data types, but I don't know how.  PyVSC documentation includes the following description:

> All PyVSC scalar attribute types provide a get_val() and set_val() method. These methods can be called to get or set the current value.

However, no example is provided.  The code shown below fails while attempting to call `get_val()` and throws the error `'ValueInt' object has no attribute 'get_val'`.  Yet I see in types.py that `get_val()` is a method of class `type_base`, which is the grandparent class of `rand_bit_t`.  I am using PyVSC version 0.6.3.

In this simple example it is not technically necessary to use `get_val()`, but in another project I genuinely need to convert PyVSC scalar data into simple integers.  How do I use these methods?

```python
import vsc

@vsc.randobj
class IntData:
    def __init__(self):
        self.number = vsc.rand_bit_t(10)


intData = IntData()
intData.randomize()

print("intData = " + intData.number.get_val())
intData.number.set_val(4)
print("intData = " + intData.number.get_val())
```
```
Traceback (most recent call last):
  File "/data/cf/sshot/u/mcgrathm/silicon-validation/experiment/pyvsc/constrDebug6.py", line 13, in <module>
    print("intData = " + intData.number.get_val())
AttributeError: 'ValueInt' object has no attribute 'get_val'
```

**Comment by mballance** [comment]
Hi @msmftc,
    As you rightly point out, this situation isn't well-covered in the docs. I suppose that is because most use cases can make use of the Python operator overloading that allows getting/setting the field value simply by referencing it. The get_val()/set_val() methods are primarily used when creating fields outside a class context:
https://github.com/fvutils/pyvsc/blob/eb6824c2c664f954cc0d8115dba67b054086c953/ve/unit/test_field_standalone.py#L22-L28

In order to call get_val/set_val directly, your code will need to work around the operator overloading. PyVSC provides something called `raw_mode` to do this. Here's an example of using `raw_mode` to access the `rand_mode` flag within a field:
https://github.com/fvutils/pyvsc/blob/3053880cb1f0caefc0f5b32c6599f400ddda29f0/ve/unit/test_rand_mode.py#L12-L67

Do keep in mind that the operator overloading calls get_val(), so I'm not positive that calling get_val() will solve your issue.
https://github.com/fvutils/pyvsc/blob/02302b461c44079eaf46d77923cb3e7e98eec920/src/vsc/rand_obj.py#L89-L108

Can you provide a bit more info on what isn't currently working right? It's possible a cast would get the desired results more-easily than directly invoking get_val.

**Comment by msmftc** [comment]
@mballance, thank-you for explaining why my calls to `get_val()` were failing.  Casting values with `int()` is a good, readable alternative.
